File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.cpp
#include <string>
#include <vector>
#include <stack>
using namespace std;

string solution(string number, int k) {
    string ans = "";
    stack<char> stk;
    int i = 0;

    while (k > 0 && i < number.size()) {
        while (!stk.empty() && k > 0 && stk.top() < number[i]) {
            stk.pop();
            k--;
        }
        stk.push(number[i]);
        i++;
    }

    while (!stk.empty()) {
        ans = stk.top() + ans;
        stk.pop();
    }
    
    ans += number.substr(i);
    if (k) ans.erase(ans.size() - 1, 1);

    return ans;
}

File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.java
import java.util.*;

class Solution {
    
    public String solution(String number, int k) {
        StringBuilder sb = new StringBuilder();

        char[] chars = number.toCharArray();
        Deque<Integer> stk = new ArrayDeque<>();
        
        stk.push(chars[0] - '0');
        int cnt = 0;
        for (int i = 1; i < chars.length; i++) {
            int val = chars[i] - '0';
            
            while (!stk.isEmpty() && stk.peek() < val && cnt < k) {
                stk.pop();
                cnt++;
            }
            
            stk.push(val);
        }       
        
        while (cnt < k) {
            stk.pop();
            cnt++;
        }
        
        while (!stk.isEmpty()) {
            sb.append(stk.pop());
        }
        
        
        return sb.reverse().toString();
    }
}
/**

**/